### PR TITLE
Adding test inference of Llama3B with prefil on device and decode on CPU

### PR DESCRIPTION
**Pull request by dgolubovicTT**
Adding test inference of Llama3B with prefil on device and decode on CPU without KV chache.

### CODE REVIEW

File: forge/test/mlir/llama/tests/test_llama_prefil.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    model_decoder.config.use_cache = False
+    model_decoder.config.output_attentions = False
+    model_decoder.config.output_hidden_states = False
+    model_decoder.config.return_dict = True
+    compiled_decoder = forge.compile(model_decoder, sample_inputs=input_ids)
+
+    # Prefill Phase - Process the initial prompt
+    transformer_outputs = compiled_decoder(input_ids)
+    # Get hidden states for all tokens from the last "transformer layer"
+    hidden_states = transformer_outputs.last_hidden_state
```
Comment (dgolubovicTT) [inline forge/test/mlir/llama/tests/test_llama_prefil.py:69]: @nvukobratTT Do we want to compare last hidden states for compiled model and torch golden? How do we want to do verification in text generation?

Comment (nvukobratTT) [reply forge/test/mlir/llama/tests/test_llama_prefil.py:69]: Do you see some benefits for comparison of hidden states? Can't we use logits (output of LM head) or just argmax out of it and compare tokens (logits are probably more flexible)?

That way, we can compile whole model with LM head as well. Let me know if I'm missing some details

File: forge/test/mlir/llama/tests/test_llama_prefil.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    model = LlamaForCausalLM.from_pretrained(model_path)
+    tokenizer = LlamaTokenizer.from_pretrained(model_path)
+
+    # Prepare input sentence
+    prompt = "Q: What is the largest animal?\nA:"
+    input_ids = tokenizer(prompt, return_tensors="pt").input_ids
+    print("length of input_ids before text generation: ", len(input_ids[0]))
+    # This is the part of the model needed for prefil; model without last Linear layer (lm_head)
+    model_decoder = model.get_decoder()
+    model_decoder.config.use_cache = False
```
Comment (nvukobratTT) [inline forge/test/mlir/llama/tests/test_llama_prefil.py:60]: Also, you can update util function for loading model with kwargs, and send these as well if needed :))

File: forge/test/mlir/llama/tests/test_llama_prefil.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+        # Stop if the EOS token is generated
+        if next_token_id == tokenizer.eos_token_id:
+            break
+
+        # Update input_ids with the new token
+        input_ids = torch.cat([input_ids, next_token_id.unsqueeze(0)], dim=-1)
+
+        # Run the model again to get the new hidden state
+        hidden_states = prefil_on_cpu(model, input_ids)
```
Comment (nvukobratTT) [inline forge/test/mlir/llama/tests/test_llama_prefil.py:36]: Let's update a naming a bit, having prefil mentioned here is a bit confusing :))

File: forge/test/mlir/llama/tests/test_llama_prefil.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    
+    return input_ids, hidden_states
```
Comment (nvukobratTT) [inline forge/test/mlir/llama/tests/test_llama_prefil.py:38]: Do we need to return hidden states after decode?

File: forge/test/mlir/llama/tests/test_llama_prefil.py (hunk 1 above)
Comment (dgolubovicTT) [reply forge/test/mlir/llama/tests/test_llama_prefil.py]: OFC that LM head is part of the model, but in this test it is ran on CPU so that was my line of resoning when I decided to compare last hidden states. But since this test will evolve so that LM is also on Device, I can switch to comparing output...

File: forge/test/mlir/llama/tests/test_llama_prefil.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+@pytest.mark.xfail()
+def test_llama_prefil_on_device_decode_on_cpu():
+    """
+    This function tests the inference of the Llama 3B model split into two parts:
+    - The first part is the prefilling of the model on the device.
+    - The second part is the decoding of the model on the CPU.
+    """
+    # Load Llama 3B model and tokenizer
+    model_path = "openlm-research/open_llama_3b"
```
[the remark this answers is not in the document]
Comment (dgolubovicTT) [reply forge/test/mlir/llama/tests/test_llama_prefil.py:48]: No, I am using this model_path to load a tokenizer since load_model doesn't return a tokenizer... Should this be inside load_model?

[the remark this answers is not in the document]
Comment (dgolubovicTT) [reply forge/test/mlir/llama/tests/test_llama_prefil.py:53]: I added loading tokenizer inside load_model method...

File: forge/test/mlir/llama/tests/test_llama_prefil.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    model = load_model(model_path, use_cache=False, return_dict=True)
+    tokenizer = LlamaTokenizer.from_pretrained(model_path)
+
+    # Prepare input sentence
+    prompt = "Q: What is the largest animal?\nA:"
+    input_ids = tokenizer(prompt, return_tensors="pt").input_ids
+    
+    # This is the part of the model needed for prefill; model without the last Linear layer (lm_head)
```
[the remark this answers is not in the document]
Comment (dgolubovicTT) [reply forge/test/mlir/llama/tests/test_llama_prefil.py:61]: As we discussed offline, prefill stage is defined as the stage where we prepare hidden states and KV cache to be able to do efficient decode. In our case, putting lm head inside prefill stage would mean we have created one token output before we started the decode. Therefore, leaving it like this because is more inline with intention of prefill and decode...

**Review by nvukobratTT — APPROVED** [review]
